README.md
# id3-js

A simple JavaScript library which can read ID3 tag from a audio file.

## Research material

Research material for how the ID3 tag works in various versions of it. Most of it is from the official website
[id3.org](https://www.id3.org/) but that seems to be down for a couple of years so I'm referencing
[web.archive.org](https://web.archive.org) instead.

- [wikipedia.org; ID3](https://en.wikipedia.org/wiki/ID3)
- [web.archive.org: id3.org/ID3v1](https://web.archive.org/web/20070930203401/http://www.id3.org/ID3v1)
- [web.archive.org: id3v2.3.0.txt](https://web.archive.org/web/20220121174529/https://id3.org/d3v2.3.0)
- [web.archive.org: id3v2.4.0-structure.txt](https://web.archive.org/web/20211112034826/https://id3.org/id3v2.4.0-structure)
- [web.archive.org: id3v2.4.0-frames.txt](https://web.archive.org/web/20211109170227/https://id3.org/id3v2.4.0-frames)
- [web.archive.org: id3v2.4.0-changes.txt](https://web.archive.org/web/20210807160919/https://id3.org/id3v2.4.0-changes)
- [web.archive.org: ID3v2 Programming Guidlines](https://web.archive.org/web/20211122000720/https://id3.org/Developer%20Information)
- [web.archive.org: id3.org: id3v1_test_suite.tar.gz](https://web.archive.org/web/20070421084950if_/http://www.id3.org:80/Developer_Information?action=AttachFile&do=get&target=id3v1_test_suite.tar.gz)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/), and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Initial code base and test
- Eslint to enforce code style
- Jest testing framework
- Rollup building system
- ID3 standards reference material
- This changelog


test/files.mjs
import fs from 'fs';

export const id3v1 = fs.readFileSync('./test/files/id3v1.mp3');
export const id3v1v2 = fs.readFileSync('./test/files/id3v1v2.mp3');
export const id3v23 = fs.readFileSync('./test/files/id3v2.3.mp3');
export const id3v24 = fs.readFileSync('./test/files/id3v2.4.mp3');
export const no_meta = fs.readFileSync('./test/files/no-meta.mp3');


test/id3.test.mjs
import * as testFile from './files.mjs';
import { readID3 } from '../src/id3.mjs';
import { expect, test } from '@jest/globals';

test('should be able to read ID3v2', () => {
	const id3v1v2 = readID3(new Uint8Array(testFile.id3v1v2.buffer));
	const id3v23 = readID3(new Uint8Array(testFile.id3v23.buffer));
	const id3v24 = readID3(new Uint8Array(testFile.id3v24.buffer));
	expect(id3v24.version).toBe('ID3v2.4.0');
	expect(id3v23.version).toBe('ID3v2.3.0');
	expect(id3v1v2.version).toBe('ID3v2.3.0');
});

test('should be able to read ID3v1', () => {
	const id3v1 = readID3(new Uint8Array(testFile.id3v1.buffer));
	expect(id3v1.version).toBe('ID3v1.0');
});

test('should return null if no ID3 tag found', () => {
	const noMeta = readID3(new Uint8Array(testFile.no_meta.buffer));
	expect(noMeta).toBeNull();
});


src/id3.mjs

import {
	readChars,
	getUInt8,
	getSyncSafeInt32,
	getSyncSafeInt35, getUInt16,
} from './binary.mjs';

import { genre } from './genre.mjs';

const readID3v1 = (file) => {
	// in ID3v1 it's always the last 128 bytes of the file
	const tag = file.slice(-128);

	// ID3v1 should always start with TAG
	if (readChars(tag, 0, 3) !== 'TAG') {
		return null;
	}

	const genreIndex = getUInt8(tag, 127);
	const id3 = {
		version: 'ID3v1.0',
		metadata: {
			title: readChars(tag, 3, 30).trim(),
			artist: readChars(tag, 33, 30).trim(),
			album: readChars(tag, 63, 30).trim(),
			year: readChars(tag, 93, 4).trim(),
			comment: readChars(tag, 97, 30).trim(),
			genre: genreIndex in genre ? genre[genreIndex] : 'Unknown',
		},
	};

	const isID3v1_1 = tag[125] === 0 && tag[126] !== 0;
	if (isID3v1_1) {
		id3.version = 'ID3v1.1';
		id3.metadata.track = getUInt8(tag, 126);
	}

	return id3;
};

const handleTextFrame = (stream) => {
	const encoding = stream[0];
	const strings = [];
	stream = stream.slice(1);
	while(stream.length) {
		const str = readChars(stream, 0, stream.length);
		strings.push(str);
		stream = stream.slice(str.length > 0 ? str.length : 1);
	}
	return strings.length === 1 ? strings[0] : strings;
};

const readID3v2 = (file) => {
	// The first 3 bytes of the file should write out 'ID3' for v2
	if (readChars(file, 0, 3) !== 'ID3') {
		return null;
	}

	// We get the size first so we can trim down to the ID3 tag only and work from that
	const size = getSyncSafeInt32(file, 6);
	let id3 = file.slice(0, size + 10);

	file = null;

	const minor = getUInt8(id3, 3);
	const patch = getUInt8(id3, 4);

	const data = {
		version: `ID3v2.${minor}.${patch}`,
		flags: {
			unsynchronisation: ((id3[5] & 0x80) >> 7) === 1,
			extendedHeader: ((id3[5] & 0x40) >> 6) === 1,
			experimentalIndicator: ((id3[5] & 0x20) >> 5) === 1,
			footerPresent: ((id3[5] & 0x10) >> 4) === 1,
		},
		size: size,
		frames: [],
	};

	id3 = id3.slice(10);

	if (data.flags.extendedHeader) {
		data.extendedHeader = {
			size: getSyncSafeInt32(id3, 0),
			numberOfFlags: getUInt8(id3, 4),
			flags: [],
		};
		// An extended header can never have a size of fewer than six bytes.
		let extendedHeaderSize = 5;
		let extendedHeader = id3.slice(5);
		for (let i = 0; i < data.extendedHeader.numberOfFlags; i++) {
			extendedHeaderSize += 1;
			const flag = extendedHeader[0];
			const flagHeader = {};
			const hasUpdate = ((flag & 0x40) >> 6) === 1;
			const hasCRC = ((flag & 0x20) >> 5) === 1;
			const hasTagRestrictions = ((flag & 0x10) >> 4) === 1;
			extendedHeader = extendedHeader.slice(1);
			flagHeader.update = hasUpdate;
			flagHeader.CRC = null;
			flagHeader.tagRestrictions = null;
			if (hasCRC) {
				flagHeader.CRC = getSyncSafeInt35(extendedHeader, 1);
				// CRC flag data length us always 5, and then we have to account for the length bit as well
				extendedHeader = extendedHeader.slice(5 + 1);
				extendedHeaderSize += 5 + 1;
			}
			if (hasTagRestrictions) {
				flagHeader.tagRestrictions = {
					tagSize: ((extendedHeader[1] & 0xc0) >> 6) & 0x3,
					textEncoding: ((extendedHeader[1] & 0x20) >> 5) === 1,
					textFieldsSize: (extendedHeader[1] & 0x18) >> 3,
					imageEncoding: ((extendedHeader[1] & 0x4) >> 2) === 1,
					imageSize: extendedHeader[1] & 0x3,
				};
				extendedHeader = extendedHeader.slice(2);
				extendedHeaderSize += 2;
			}
			data.extendedHeader.flags.push(flagHeader);
		}
		// Let's throw a fatal error if the reported header size does not match up with out calculations
		if (data.extendedHeader.size !== extendedHeaderSize) {
			throw 'Unexpected header size in ID3v2 extended header';
		}
		id3 = extendedHeader;
	}

	while (/^[A-Z0-9]{4}$/.test(readChars(id3, 0, 4))) {
		const frameSize = getSyncSafeInt32(id3, 4);
		const frame = {
			id: readChars(id3, 0, 4),
			size: frameSize,
			flags: {
				tagAlterPreservation: ((id3[8] & 0x40) >> 6) === 1,
				fileAlterPreservation: ((id3[8] & 0x20) >> 5) === 1,
				readOnly: ((id3[8] & 0x10) >> 4) === 1,
				groupingIdentity: ((id3[9] & 0x40) >> 6) === 1,
				compression: ((id3[9] & 0x8) >> 3) === 1,
				encryption: ((id3[9] & 0x4) >> 2) === 1,
				unsynchronisation: ((id3[9] & 0x2) >> 1) === 1,
				dataLengthIndicator: (id3[9] & 0x1) === 1,
			},
			data: id3.slice(10, frameSize + 10),
		};
		if (frame.id[0] === 'T' && frame.id !== 'TXXX') {
			frame.data = handleTextFrame(frame.data);
		}
		data.frames.push(frame);
		id3 = id3.slice(frame.size + 10);
	}

	return data;
};

export const readID3 = (file) => {

	const id3v2 = readID3v2(file);
	if (id3v2) {
		return id3v2;
	}
	return readID3v1(file);
};


src/binary.mjs

const readChars = (uint8Array, position, length) => {
	if (!(uint8Array instanceof Uint8Array)) { throw new TypeError('uint8Array must an instance of Uint8Array'); }
	if (!Number.isInteger(position)) { throw new TypeError('position must an integer'); }
	if (!Number.isInteger(length)) { throw new TypeError('length must an integer'); }

	let data = [...uint8Array.slice(position, position + length)];

	const nullByte = data.findIndex((byte) => byte === 0);
	if (nullByte !== -1) {
		data = [...data.slice(0, nullByte)];
	}

	return String.fromCharCode(...data);
};

const findMarker = (haystack, needle) => {
	if (!(haystack instanceof Uint8Array)) { throw new TypeError('haystack must an instance of Uint8Array'); }

	const marker = strToUint8Array(needle);
	const match = marker.toString();
	for (let i = 0; i < haystack.length; i++) {
		if (haystack.slice(i, i + marker.length).toString() === match) {
			return i;
		}
	}
	return null;
};

const findMarkers = (haystack, needle) => {
	if (!(haystack instanceof Uint8Array)) { throw new TypeError('haystack must an instance of Uint8Array'); }

	const marker = strToUint8Array(needle);
	const match = marker.toString();
	const markerPositions = [];
	for (let i = 0; i < haystack.length; i++) {
		if (haystack.slice(i, i + marker.length).toString() === match) {
			markerPositions.push(i);
		}
	}
	return markerPositions;
};

const strToUint8Array = (str) => {
	const chars = str.split('').map((letter) => letter.charCodeAt(0));
	return new Uint8Array(chars);
};

const getFixedPoint32 = (uint8Array, byteOffset) => {
	if (!Number.isInteger(byteOffset)) { throw new TypeError('Byte offset must be an integer'); }

	const first = (new DataView(uint8Array.buffer)).getUint16(byteOffset, false);
	const last = (new DataView(uint8Array.buffer)).getUint16(byteOffset + 2, false);
	return parseFloat(`${first}.${last}`);
};

const getFixedPoint16 = (uint8Array, byteOffset) => {
	if (!(uint8Array instanceof Uint8Array)) { throw new TypeError('uint8Array must an instance of Uint8Array'); }
	if (!Number.isInteger(byteOffset)) { throw new TypeError('Byte offset must be an integer'); }

	const first = uint8Array[byteOffset];
	const last = uint8Array[byteOffset + 1];
	return parseFloat(`${first}.${last}`);
};

const getInt8 = (uint8Array, byteOffset) => {
	if (!Number.isInteger(byteOffset)) { throw new TypeError('Byte offset must be an integer'); }

	return (new DataView(uint8Array.buffer))
		.getInt8(byteOffset);
};

const getUInt8 = (uint8Array, byteOffset) => {
	if (!Number.isInteger(byteOffset)) { throw new TypeError('Byte offset must be an integer'); }

	return (new DataView(uint8Array.buffer))
		.getUint8(byteOffset);
};

const getInt16 = (uint8Array, byteOffset) => {
	if (!Number.isInteger(byteOffset)) { throw new TypeError('Byte offset must be an integer'); }

	return (new DataView(uint8Array.buffer))
		.getInt16(byteOffset, false);
};

const getUInt16 = (uint8Array, byteOffset) => {
	if (!Number.isInteger(byteOffset)) { throw new TypeError('Byte offset must be an integer'); }

	return (new DataView(uint8Array.buffer))
		.getUint16(byteOffset, false);
};

const getInt32 = (uint8Array, byteOffset) => {
	if (!Number.isInteger(byteOffset)) { throw new TypeError('Byte offset must be an integer'); }

	return (new DataView(uint8Array.buffer))
		.getInt32(byteOffset, false);
};

const getUInt32 = (uint8Array, byteOffset) => {
	if (!Number.isInteger(byteOffset)) { throw new TypeError('Byte offset must be an integer'); }

	return (new DataView(uint8Array.buffer))
		.getUint32(byteOffset, false);
};

// Most integers in ID3v2 spec are stored as "syncsafe" integers for some reason.
// Inspiration: https://stackoverflow.com/questions/7898991/why-sync-safe-integer
const getSyncSafeInt32 = (uint8Array, byteOffset) => {
	const integer = getUInt32(uint8Array, byteOffset);
	return (integer & 0x7f) |
		(integer & 0x7f00) >> 1 |
		(integer & 0x7f0000) >> 2 |
		(integer & 0x7f000000) >> 3;
};

const getSyncSafeInt35 = (uint8Array, byteOffset) => {
	const data = new Uint8Array([...uint8Array.slice(byteOffset, byteOffset + 5)]);
	const last =  ((data[3] & 0x1) << 7) | data[4];
	const nr3 =  ((data[2] & 0x3) << 6) | (data[3] >> 1);
	const nr2 =  ((data[1] & 0x7) << 5) | (data[2] >> 2);
	const nr1 =  (data[0] << 4) | (data[1] >> 3);
	return getUInt32(new Uint8Array([nr1, nr2, nr3, last]), 0);
};

export {
	readChars,
	findMarker,
	findMarkers,
	strToUint8Array,
	getInt8,
	getInt16,
	getInt32,
	getUInt8,
	getUInt16,
	getUInt32,
	getFixedPoint16,
	getFixedPoint32,
	getSyncSafeInt32,
	getSyncSafeInt35,
};


src/genre.mjs

export const genre = {
	0: 'Blues',
	1: 'Classic Rock',
	2: 'Country',
	3: 'Dance',
	4: 'Disco',
	5: 'Funk',
	6: 'Grunge',
	7: 'Hip-Hop',
	8: 'Jazz',
	9: 'Metal',
	10: 'New Age',
	11: 'Oldies',
	12: 'Other',
	13: 'Pop',
	14: 'R&B',
	15: 'Rap',
	16: 'Reggae',
	17: 'Rock',
	18: 'Techno',
	19: 'Industrial',
	20: 'Alternative',
	21: 'Ska',
	22: 'Death Metal',
	23: 'Pranks',
	24: 'Soundtrack',
	25: 'Euro-Techno',
	26: 'Ambient',
	27: 'Trip-Hop',
	28: 'Vocal',
	29: 'Jazz+Funk',
	30: 'Fusion',
	31: 'Trance',
	32: 'Classical',
	33: 'Instrumental',
	34: 'Acid',
	35: 'House',
	36: 'Game',
	37: 'Sound Clip',
	38: 'Gospel',
	39: 'Noise',
	40: 'AlternRock',
	41: 'Bass',
	42: 'Soul',
	43: 'Punk',
	44: 'Space',
	45: 'Meditative',
	46: 'Instrumental Pop',
	47: 'Instrumental Rock',
	48: 'Ethnic',
	49: 'Gothic',
	50: 'Darkwave',
	51: 'Techno-Industrial',
	52: 'Electronic',
	53: 'Pop-Folk',
	54: 'Eurodance',
	55: 'Dream',
	56: 'Southern Rock',
	57: 'Comedy',
	58: 'Cult',
	59: 'Gangsta',
	60: 'Top 40',
	61: 'Christian Rap',
	62: 'Pop/Funk',
	63: 'Jungle',
	64: 'Native American',
	65: 'Cabaret',
	66: 'New Wave',
	67: 'Psychadelic',
	68: 'Rave',
	69: 'Showtunes',
	70: 'Trailer',
	71: 'Lo-Fi',
	72: 'Tribal',
	73: 'Acid Punk',
	74: 'Acid Jazz',
	75: 'Polka',
	76: 'Retro',
	77: 'Musical',
	78: 'Rock & Roll',
	79: 'Hard Rock',
	80: 'Folk',
	81: 'Folk-Rock',
	82: 'National Folk',
	83: 'Swing',
	84: 'Fast Fusion',
	85: 'Bebob',
	86: 'Latin',
	87: 'Revival',
	88: 'Celtic',
	89: 'Bluegrass',
	90: 'Avantgarde',
	91: 'Gothic Rock',
	92: 'Progressive Rock',
	93: 'Psychedelic Rock',
	94: 'Symphonic Rock',
	95: 'Slow Rock',
	96: 'Big Band',
	97: 'Chorus',
	98: 'Easy Listening',
	99: 'Acoustic',
	100: 'Humour',
	101: 'Speech',
	102: 'Chanson',
	103: 'Opera',
	104: 'Chamber Music',
	105: 'Sonata',
	106: 'Symphony',
	107: 'Booty Bass',
	108: 'Primus',
	109: 'Porn Groove',
	110: 'Satire',
	111: 'Slow Jam',
	112: 'Club',
	113: 'Tango',
	114: 'Samba',
	115: 'Folklore',
	116: 'Ballad',
	117: 'Power Ballad',
	118: 'Rhythmic Soul',
	119: 'Freestyle',
	120: 'Duet',
	121: 'Punk Rock',
	122: 'Drum Solo',
	123: 'A capella',
	124: 'Euro-House',
	125: 'Dance Hall',
	126: 'Goa',
	127: 'Drum & Bass',
	128: 'Club-House',
	129: 'Hardcore',
	130: 'Terror',
	131: 'Indie',
	132: 'BritPop',
	133: 'Negerpunk',
	134: 'Polsk Punk',
	135: 'Beat',
	136: 'Christian',
	137: 'Heavy Metal',
	138: 'Black Metal',
	139: 'Crossover',
	140: 'Contemporary',
	141: 'Christian Rock',
	142: 'Merengue',
	143: 'Salsa',
	144: 'Thrash Metal',
	145: 'Anime',
	146: 'JPop',
	147: 'Synthpop',
};
